### PR TITLE
Approve login link workflow

### PR DESCRIPTION
**Pull request by carmenlau**
refs #2906

### CODE REVIEW

**Comment by kiootic** [comment]
Some issues encountered while trying to integrate this feature:
- Need to add workflow ID to the OTP code (like WebSessionID), but nodes in sub-workflow does not have correct workflow ID
- `NodeAuthenticateEmailLoginLink` does not have output indicating whether a code has been accepted by server; so either new a workflow event asking client to submit `latte.InputCheckLoginLinkVerified`, or add a flag to node output indicating it.
- #2934 makes CORS WebSocket not work.

**Comment by kiootic** [comment]
My quick patch for your reference:
```diff
diff --git a/pkg/auth/deps.go b/pkg/auth/deps.go
index c0a47bad7..22547eaec 100644
--- a/pkg/auth/deps.go
+++ b/pkg/auth/deps.go
@@ -126,7 +126,7 @@ var DependencySet = wire.NewSet(
 	wire.Bind(new(handlerapi.WorkflowNewWorkflowService), new(*workflow.Service)),
 	wire.Bind(new(handlerapi.WorkflowGetWorkflowService), new(*workflow.Service)),
 	wire.Bind(new(handlerapi.WorkflowInputWorkflowService), new(*workflow.Service)),
-	wire.Bind(new(handlerapi.WorkflowWebsocketEventStore), new(*workflow.EventStore)),
+	wire.Bind(new(handlerapi.WorkflowWebsocketEventStore), new(*workflow.EventStoreImpl)),
 
 	viewmodelswebapp.DependencySet,
 	wire.Bind(new(viewmodelswebapp.StaticAssetResolver), new(*web.StaticAssetResolver)),
diff --git a/pkg/auth/handler/api/workflow_websocket.go b/pkg/auth/handler/api/workflow_websocket.go
index f5100ed88..3e98eb10c 100644
--- a/pkg/auth/handler/api/workflow_websocket.go
+++ b/pkg/auth/handler/api/workflow_websocket.go
@@ -27,9 +27,10 @@ type WorkflowWebsocketHandler struct {
 
 func (h *WorkflowWebsocketHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	handler := &pubsub.HTTPHandler{
-		RedisHub:      h.RedisHandle,
-		Delegate:      h,
-		LoggerFactory: h.LoggerFactory,
+		InsecureSkipVerify: true,
+		RedisHub:           h.RedisHandle,
+		Delegate:           h,
+		LoggerFactory:      h.LoggerFactory,
 	}
 
 	handler.ServeHTTP(w, r)
diff --git a/pkg/auth/wire_gen.go b/pkg/auth/wire_gen.go
index b897451e9..613ca3b17 100644
--- a/pkg/auth/wire_gen.go
+++ b/pkg/auth/wire_gen.go
@@ -52785,6 +52785,12 @@ func newAPIWorkflowNewHandler(p *deps.RequestProvider) http.Handler {
 		SMSConfig:   smsConfig,
 		EmailConfig: emailConfig,
 	}
+	workflowStoreImpl := &workflow.StoreImpl{
+		Redis:   appredisHandle,
+		AppID:   appID,
+		Context: contextContext,
+	}
+	eventStoreImpl := workflow.NewEventStore(appID, appredisHandle, workflowStoreImpl)
 	dependencies := &workflow.Dependencies{
 		Config:                appConfig,
 		FeatureConfig:         featureConfig,
@@ -52805,16 +52811,12 @@ func newAPIWorkflowNewHandler(p *deps.RequestProvider) http.Handler {
 		Events:                eventService,
 		RateLimiter:           limiter,
 		AntiSpamOTPCodeBucket: antiSpamOTPCodeBucketMaker,
+		WorkflowEvents:        eventStoreImpl,
 	}
 	serviceLogger := workflow.NewServiceLogger(factory)
 	savePointImpl := &workflow.SavePointImpl{
 		SQLExecutor: sqlExecutor,
 	}
-	workflowStoreImpl := &workflow.StoreImpl{
-		Redis:   appredisHandle,
-		AppID:   appID,
-		Context: contextContext,
-	}
 	workflowService := &workflow.Service{
 		ContextDoNotUseDirectly: contextContext,
 		Deps:                    dependencies,
@@ -53468,6 +53470,12 @@ func newAPIWorkflowGetHandler(p *deps.RequestProvider) http.Handler {
 		SMSConfig:   smsConfig,
 		EmailConfig: emailConfig,
 	}
+	workflowStoreImpl := &workflow.StoreImpl{
+		Redis:   appredisHandle,
+		AppID:   appID,
+		Context: contextContext,
+	}
+	eventStoreImpl := workflow.NewEventStore(appID, appredisHandle, workflowStoreImpl)
 	dependencies := &workflow.Dependencies{
 		Config:                appConfig,
 		FeatureConfig:         featureConfig,
@@ -53488,16 +53496,12 @@ func newAPIWorkflowGetHandler(p *deps.RequestProvider) http.Handler {
 		Events:                eventService,
 		RateLimiter:           limiter,
 		AntiSpamOTPCodeBucket: antiSpamOTPCodeBucketMaker,
+		WorkflowEvents:        eventStoreImpl,
 	}
 	serviceLogger := workflow.NewServiceLogger(factory)
 	savePointImpl := &workflow.SavePointImpl{
 		SQLExecutor: sqlExecutor,
 	}
-	workflowStoreImpl := &workflow.StoreImpl{
-		Redis:   appredisHandle,
-		AppID:   appID,
-		Context: contextContext,
-	}
 	workflowService := &workflow.Service{
 		ContextDoNotUseDirectly: contextContext,
 		Deps:                    dependencies,
@@ -54121,6 +54125,12 @@ func newAPIWorkflowInputHandler(p *deps.RequestProvider) http.Handler {
 		SMSConfig:   smsConfig,
 		EmailConfig: emailConfig,
 	}
+	workflowStoreImpl := &workflow.StoreImpl{
+		Redis:   appredisHandle,
+		AppID:   appID,
+		Context: contextContext,
+	}
+	eventStoreImpl := workflow.NewEventStore(appID, appredisHandle, workflowStoreImpl)
 	dependencies := &workflow.Dependencies{
 		Config:                appConfig,
 		FeatureConfig:         featureConfig,
@@ -54141,16 +54151,12 @@ func newAPIWorkflowInputHandler(p *deps.RequestProvider) http.Handler {
 		Events:                eventService,
 		RateLimiter:           limiter,
 		AntiSpamOTPCodeBucket: antiSpamOTPCodeBucketMaker,
+		WorkflowEvents:        eventStoreImpl,
 	}
 	serviceLogger := workflow.NewServiceLogger(factory)
 	savePointImpl := &workflow.SavePointImpl{
 		SQLExecutor: sqlExecutor,
 	}
-	workflowStoreImpl := &workflow.StoreImpl{
-		Redis:   appredisHandle,
-		AppID:   appID,
-		Context: contextContext,
-	}
 	workflowService := &workflow.Service{
 		ContextDoNotUseDirectly: contextContext,
 		Deps:                    dependencies,
@@ -54180,10 +54186,10 @@ func newAPIWorkflowWebsocketHandler(p *deps.RequestProvider) http.Handler {
 		AppID:   appID,
 		Context: contextContext,
 	}
-	eventStore := workflow.NewEventStore(appID, handle, storeImpl)
+	eventStoreImpl := workflow.NewEventStore(appID, handle, storeImpl)
 	factory := appProvider.LoggerFactory
 	workflowWebsocketHandler := &api.WorkflowWebsocketHandler{
-		Events:        eventStore,
+		Events:        eventStoreImpl,
 		LoggerFactory: factory,
 		RedisHandle:   handle,
 	}
diff --git a/pkg/latte/intent_authenticate_email_login_link.go b/pkg/latte/intent_authenticate_email_login_link.go
index 2fbf35f88..ff31ba2fa 100644
--- a/pkg/latte/intent_authenticate_email_login_link.go
+++ b/pkg/latte/intent_authenticate_email_login_link.go
@@ -49,6 +49,7 @@ func (i *IntentAuthenticateEmailLoginLink) ReactTo(ctx context.Context, deps *wo
 			IsAuthenticating:  true,
 			AuthenticatorInfo: authenticator,
 			OTPMode:           otp.OTPModeMagicLink,
+			WorkflowID:        w.WorkflowID,
 		}).Do()
 		if apierrors.IsKind(err, ratelimit.RateLimited) {
 			// Ignore rate limit error for initial sending
diff --git a/pkg/latte/intent_authenticate_oob_otp_phone.go b/pkg/latte/intent_authenticate_oob_otp_phone.go
index d2da7710c..58a8ee3c3 100644
--- a/pkg/latte/intent_authenticate_oob_otp_phone.go
+++ b/pkg/latte/intent_authenticate_oob_otp_phone.go
@@ -47,6 +47,7 @@ func (i *IntentAuthenticateOOBOTPPhone) ReactTo(ctx context.Context, deps *workf
 			Stage:             authenticatorKindToStage(authenticator.Kind),
 			IsAuthenticating:  true,
 			AuthenticatorInfo: authenticator,
+			WorkflowID:        w.WorkflowID,
 		}).Do()
 		if apierrors.IsKind(err, ratelimit.RateLimited) {
 			// Ignore rate limit error for initial sending
diff --git a/pkg/latte/intent_verify_login_link.go b/pkg/latte/intent_verify_login_link.go
index e05749dbb..e41fb4ab8 100644
--- a/pkg/latte/intent_verify_login_link.go
+++ b/pkg/latte/intent_verify_login_link.go
@@ -39,13 +39,15 @@ func (i *IntentVerifyLoginLink) ReactTo(ctx context.Context, deps *workflow.Depe
 	switch {
 	case workflow.AsInput(input, &inputTakeLoginLinkCode):
 		code := inputTakeLoginLinkCode.GetCode()
-		_, err := deps.OTPCodes.SetUserInputtedMagicLinkCode(code)
+		otp, err := deps.OTPCodes.SetUserInputtedMagicLinkCode(code)
 		if err != nil {
 			return nil, err
 		}
-		// TODO(workflow): Send websocket event for refreshing workflow
+
 		return workflow.NewNodeSimple(
-			&NodeVerifiedLoginLink{},
+			&NodeVerifiedLoginLink{
+				OriginWorkflowID: otp.WorkflowID,
+			},
 		), nil
 	default:
 		return nil, workflow.ErrIncompatibleInput
diff --git a/pkg/latte/node_authenticate_email_login_link.go b/pkg/latte/node_authenticate_email_login_link.go
index a6f20129f..b91c0c094 100644
--- a/pkg/latte/node_authenticate_email_login_link.go
+++ b/pkg/latte/node_authenticate_email_login_link.go
@@ -45,6 +45,7 @@ func (n *NodeAuthenticateEmailLoginLink) ReactTo(ctx context.Context, deps *work
 			IsAuthenticating:  true,
 			AuthenticatorInfo: info,
 			OTPMode:           otp.OTPModeMagicLink,
+			WorkflowID:        w.WorkflowID,
 		}).Do()
 		if err != nil {
 			return nil, err
diff --git a/pkg/latte/node_authenticate_oob_otp_phone.go b/pkg/latte/node_authenticate_oob_otp_phone.go
index e34023692..7e8ab5acb 100644
--- a/pkg/latte/node_authenticate_oob_otp_phone.go
+++ b/pkg/latte/node_authenticate_oob_otp_phone.go
@@ -43,6 +43,7 @@ func (n *NodeAuthenticateOOBOTPPhone) ReactTo(ctx context.Context, deps *workflo
 			Stage:             authenticatorKindToStage(info.Kind),
 			IsAuthenticating:  true,
 			AuthenticatorInfo: info,
+			WorkflowID:        w.WorkflowID,
 		}).Do()
 		if err != nil {
 			return nil, err
diff --git a/pkg/latte/node_verified_login_link.go b/pkg/latte/node_verified_login_link.go
index ba729dab8..4ae48466b 100644
--- a/pkg/latte/node_verified_login_link.go
+++ b/pkg/latte/node_verified_login_link.go
@@ -11,6 +11,7 @@ func init() {
 }
 
 type NodeVerifiedLoginLink struct {
+	OriginWorkflowID string `json:"origin_workflow_id"`
 }
 
 func (n *NodeVerifiedLoginLink) Kind() string {
@@ -18,7 +19,11 @@ func (n *NodeVerifiedLoginLink) Kind() string {
 }
 
 func (n *NodeVerifiedLoginLink) GetEffects(ctx context.Context, deps *workflow.Dependencies, w *workflow.Workflow) (effs []workflow.Effect, err error) {
-	return nil, nil
+	return []workflow.Effect{
+		workflow.OnCommitEffect(func(ctx context.Context, deps *workflow.Dependencies) error {
+			return deps.WorkflowEvents.Publish(n.OriginWorkflowID, workflow.NewEventMagicLinkCodeVerified())
+		}),
+	}, nil
 }
 
 func (n *NodeVerifiedLoginLink) CanReactTo(ctx context.Context, deps *workflow.Dependencies, w *workflow.Workflow) ([]workflow.Input, error) {
diff --git a/pkg/latte/node_verify_email.go b/pkg/latte/node_verify_email.go
index 8e038f52e..7a289e6f9 100644
--- a/pkg/latte/node_verify_email.go
+++ b/pkg/latte/node_verify_email.go
@@ -109,8 +109,7 @@ func (n *NodeVerifyEmail) sendCode(deps *workflow.Dependencies, w *workflow.Work
 		return err
 	}
 
-	// FIXME: web session ID?
-	code, err := deps.OTPCodes.GenerateCode(n.Email, otp.OTPModeCode, string(deps.Config.ID), "")
+	code, err := deps.OTPCodes.GenerateCode(n.Email, otp.OTPModeCode, string(deps.Config.ID), w.WorkflowID)
 	if err != nil {
 		return err
 	}
diff --git a/pkg/latte/node_verify_phone_sms.go b/pkg/latte/node_verify_phone_sms.go
index 3e2540210..91329682c 100644
--- a/pkg/latte/node_verify_phone_sms.go
+++ b/pkg/latte/node_verify_phone_sms.go
@@ -117,7 +117,7 @@ func (n *NodeVerifyPhoneSMS) sendCode(deps *workflow.Dependencies, w *workflow.W
 	}
 
 	// FIXME: web session ID?
-	code, err := deps.OTPCodes.GenerateCode(n.PhoneNumber, otp.OTPModeCode, string(deps.Config.ID), "")
+	code, err := deps.OTPCodes.GenerateCode(n.PhoneNumber, otp.OTPModeCode, string(deps.Config.ID), w.WorkflowID)
 	if err != nil {
 		return err
 	}
diff --git a/pkg/latte/utils_authn.go b/pkg/latte/utils_authn.go
index 50d156e1b..621a3cc7d 100644
--- a/pkg/latte/utils_authn.go
+++ b/pkg/latte/utils_authn.go
@@ -15,6 +15,7 @@ type SendOOBCode struct {
 	IsAuthenticating  bool
 	AuthenticatorInfo *authenticator.Info
 	OTPMode           otp.OTPMode
+	WorkflowID        string
 }
 
 func (p *SendOOBCode) Do() (*otp.CodeSendResult, error) {
@@ -70,8 +71,7 @@ func (p *SendOOBCode) Do() (*otp.CodeSendResult, error) {
 		return nil, err
 	}
 
-	// fixme(workflow): update web session id for magic link
-	code, err := p.Deps.OTPCodes.GenerateCode(p.AuthenticatorInfo.OOBOTP.ToTarget(), p.OTPMode, string(p.Deps.Config.ID), "")
+	code, err := p.Deps.OTPCodes.GenerateCode(p.AuthenticatorInfo.OOBOTP.ToTarget(), p.OTPMode, string(p.Deps.Config.ID), p.WorkflowID)
 	if err != nil {
 		return nil, err
 	}
diff --git a/pkg/lib/authn/otp/code.go b/pkg/lib/authn/otp/code.go
index d39d89f56..e389d5a89 100644
--- a/pkg/lib/authn/otp/code.go
+++ b/pkg/lib/authn/otp/code.go
@@ -18,4 +18,5 @@ type Code struct {
 	UserInputtedCode string `json:"user_inputted_code,omitempty"`
 	AppID            string `json:"app_id,omitempty"`
 	WebSessionID     string `json:"web_session_id,omitempty"`
+	WorkflowID       string `json:"workflow_id,omitempty"`
 }
diff --git a/pkg/lib/authn/otp/service.go b/pkg/lib/authn/otp/service.go
index 94b0977e2..fa4ba897c 100644
--- a/pkg/lib/authn/otp/service.go
+++ b/pkg/lib/authn/otp/service.go
@@ -114,6 +114,7 @@ func (s *Service) GenerateCode(target string, otpMode OTPMode, appID string, web
 	return s.createCode(target, otpMode, &Code{
 		AppID:        appID,
 		WebSessionID: webSessionID,
+		WorkflowID:   webSessionID, // FIXME: use workflow instead of web session
 	})
 }
 
diff --git a/pkg/lib/workflow/dependencies.go b/pkg/lib/workflow/dependencies.go
index e1332303d..f1ceed7c2 100644
--- a/pkg/lib/workflow/dependencies.go
+++ b/pkg/lib/workflow/dependencies.go
@@ -102,6 +102,10 @@ type AntiSpamOTPCodeBucketMaker interface {
 	MakeBucket(channel model.AuthenticatorOOBChannel, target string) ratelimit.Bucket
 }
 
+type EventStore interface {
+	Publish(workflowID string, e Event) error
+}
+
 type Dependencies struct {
 	Config        *config.AppConfig
 	FeatureConfig *config.FeatureConfig
@@ -127,4 +131,5 @@ type Dependencies struct {
 	Events                EventService
 	RateLimiter           RateLimiter
 	AntiSpamOTPCodeBucket AntiSpamOTPCodeBucketMaker
+	WorkflowEvents        EventStore
 }
diff --git a/pkg/lib/workflow/deps.go b/pkg/lib/workflow/deps.go
index 08741ffe9..3ad469f86 100644
--- a/pkg/lib/workflow/deps.go
+++ b/pkg/lib/workflow/deps.go
@@ -12,5 +12,6 @@ var DependencySet = wire.NewSet(
 	wire.Struct(new(Service), "*"),
 	wire.Struct(new(StoreImpl), "*"),
 	wire.Bind(new(Store), new(*StoreImpl)),
+	wire.Bind(new(EventStore), new(*EventStoreImpl)),
 	NewEventStore,
 )
diff --git a/pkg/lib/workflow/event.go b/pkg/lib/workflow/event.go
index a813f5c2b..22c72b3bc 100644
--- a/pkg/lib/workflow/event.go
+++ b/pkg/lib/workflow/event.go
@@ -5,6 +5,8 @@ type EventKind string
 const (
 	// WorkflowEventKindRefresh indicates client should re-fetch current instance of workflow for updated state.
 	EventKindRefresh EventKind = "refresh"
+	// WorkflowEventKindMagicLinkCodeVerified indicates client should proceed since magic link code is verified by server.
+	WorkflowEventKindMagicLinkCodeVerified EventKind = "magic-link-code-verified"
 )
 
 type Event interface {
@@ -22,3 +24,15 @@ func NewEventRefresh() *EventRefresh {
 func (*EventRefresh) kind() EventKind { return EventKindRefresh }
 
 var _ Event = &EventRefresh{}
+
+type EventMagicLinkCodeVerified struct {
+	Kind EventKind `json:"kind"`
+}
+
+func NewEventMagicLinkCodeVerified() *EventMagicLinkCodeVerified {
+	return &EventMagicLinkCodeVerified{Kind: WorkflowEventKindMagicLinkCodeVerified}
+}
+
+func (*EventMagicLinkCodeVerified) kind() EventKind { return WorkflowEventKindMagicLinkCodeVerified }
+
+var _ Event = &EventMagicLinkCodeVerified{}
diff --git a/pkg/lib/workflow/event_store.go b/pkg/lib/workflow/event_store.go
index dee5954b8..2a2f6d16c 100644
--- a/pkg/lib/workflow/event_store.go
+++ b/pkg/lib/workflow/event_store.go
@@ -17,15 +17,15 @@ func (p eventRedisPool) Get() *goredis.Client {
 	return p.Handle.Client()
 }
 
-type EventStore struct {
+type EventStoreImpl struct {
 	AppID       config.AppID
 	RedisHandle *appredis.Handle
 	Store       Store
 	publisher   *pubsub.Publisher
 }
 
-func NewEventStore(appID config.AppID, handle *appredis.Handle, store Store) *EventStore {
-	s := &EventStore{
+func NewEventStore(appID config.AppID, handle *appredis.Handle, store Store) *EventStoreImpl {
+	s := &EventStoreImpl{
 		AppID:       appID,
 		RedisHandle: handle,
 		Store:       store,
@@ -34,7 +34,7 @@ func NewEventStore(appID config.AppID, handle *appredis.Handle, store Store) *Ev
 	return s
 }
 
-func (s *EventStore) Publish(workflowID string, e Event) error {
+func (s *EventStoreImpl) Publish(workflowID string, e Event) error {
 	channelName, err := s.ChannelName(workflowID)
 	if errors.Is(err, ErrWorkflowNotFound) {
 		// Treat events to an non-existent (e.g. expired) workflow as noop.
@@ -56,7 +56,7 @@ func (s *EventStore) Publish(workflowID string, e Event) error {
 	return nil
 }
 
-func (s *EventStore) ChannelName(workflowID string) (string, error) {
+func (s *EventStoreImpl) ChannelName(workflowID string) (string, error) {
 	// Ignore events for workflows without session.
 	_, err := s.Store.GetSession(workflowID)
 	if err != nil {
diff --git a/pkg/lib/workflow/workflow.go b/pkg/lib/workflow/workflow.go
index 15a0a2b13..3c52152b3 100644
--- a/pkg/lib/workflow/workflow.go
+++ b/pkg/lib/workflow/workflow.go
@@ -69,6 +69,11 @@ func (w *Workflow) Accept(ctx context.Context, deps *Dependencies, input Input)
 		var nextNode *Node
 		nextNode, err = inputReactor.ReactTo(ctx, deps, workflowInQuestion, input)
 
+		// HACK: inject workflow ID to propagate it through entire workflow graph
+		if nextNode != nil && nextNode.SubWorkflow != nil {
+			nextNode.SubWorkflow.WorkflowID = workflowInQuestion.WorkflowID
+		}
+
 		// Handle err == ErrIncompatibleInput
 		if errors.Is(err, ErrIncompatibleInput) {
 			err = nil
diff --git a/pkg/util/pubsub/http_handler.go b/pkg/util/pubsub/http_handler.go
index 61a4d5c59..6cb3ffc93 100644
--- a/pkg/util/pubsub/http_handler.go
+++ b/pkg/util/pubsub/http_handler.go
@@ -34,9 +34,10 @@ const (
 // For each websocket connection, a redis pubsub connection is established.
 // Every message from the redis pubsub connection is forwarded to the websocket connection verbatim.
 type HTTPHandler struct {
-	RedisHub      RedisHub
-	Delegate      Delegate
-	LoggerFactory *log.Factory
+	InsecureSkipVerify bool
+	RedisHub           RedisHub
+	Delegate           Delegate
+	LoggerFactory      *log.Factory
 }
 
 func (h *HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
@@ -53,7 +54,9 @@ func (h *HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		logger.Debug("canceled root context")
 	}()
 
-	wsConn, err := websocket.Accept(w, r, &websocket.AcceptOptions{})
+	wsConn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		InsecureSkipVerify: h.InsecureSkipVerify,
+	})
 	if err != nil {
 		logger.WithError(err).Debug("failed to accept websocket connection")
 		return

```

**Comment by carmenlau** [comment]
@kiootic Updated accordingly, see if it works for you!

**Comment by kiootic** [comment]
Thanks it works! Not sure if we should put workflow ID in context though.

**Comment by carmenlau** [comment]
@kiootic Discuss with @louischan-oursky, we should not introduce a new event `EventMagicLinkCodeVerified`. Instead, we should fire the refresh event.

1. The node output should have a flag to indicate the code is submitted. 
2. When the web app receives the refresh event, it should reload to get the workflow to check the flag in step 1.
3. Submit `latte.InputCheckLoginLinkVerified` when the flag is true.
4. The web app may miss the refresh event. The web app should reload the page after reconnecting the websocket and run steps 2-3.

In this way, we can handle both receiving refresh events and reconnecting in the same way. I will update the PR accordingly, and see if this approach looks good for you?

File: pkg/lib/workflow/event.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ type EventKind string
 const (
 	// WorkflowEventKindRefresh indicates client should re-fetch current instance of workflow for updated state.
 	EventKindRefresh EventKind = "refresh"
+	// WorkflowEventKindLoginLinkCodeVerified indicates client should proceed since login link code is verified by server.
```
[the remark this answers is not in the document]
Comment (carmenlau) [reply pkg/lib/workflow/event.go]: Updated!

**Pull request merged**